### ISSUE TITLE
Trouble running update definition with OpenCL target

### ISSUE THREAD

**Issue by jaakkotuosa** [halide/Halide#608]
Hi!
If I add [https://github.com/jaakkotuosa/Halide/compare/gpu_update_scheduling?expand=1](this update definition) to the lesson 12 I get get `CL: clEnqueueNDRangeKernel failed: CL_OUT_OF_RESOURCES`in `lesson_12_using_the_gpu.cpp:212`.

I looked for [https://www.diffchecker.com/0ji3bz0j](differences in debug output (original on the left, added update definition on the right), and noticed that the allocation of, and access to, shared memory seems to change. That could explaint the error. Trying to understand the generated code, I would guess it is trying to allocate shared memory for all the image, not just (8+2)*(8+2)*3 bytes needed for one block, and is trying to perform the update definition (line 22 in diff) in every block.

What do you think, is the generated to code ok? If not, what should be changed? The update definition I made up is one obvious candidate, it may well be nonsensical somehow.  

I executed the code with `LD_LIBRARY_PATH=../bin HL_TARGET=opencl HL_OCL_DEVICE_TYPE=gpu HL_DEBUG_CODEGEN=1 ./lesson_12`, on Ubuntu 14.04.1 LTS, 64 bit, Radeon R9 290 4GB with fglrx 2:14.501-0ubuntu1.

**Comment by dsharletg** [comment]
You're most likely correct, the problem is that it tries to execute the
update in every block, which causes it to allocate space for the image of
the union of the block and the point 0, 0, 0, the upper bound of which is
the whole image.

BTW, if you use the target HL_TARGET=opencl-debug, you'll get debug
information for how clEnqueueNDRangeKernel is called, including the tile
size and the amount of shared/local memory requested. It should be clear if
this is actually the problem in that output.

If you need this actual thing to happen (setting the global coordinate 0,
0, 0 to 0), I see the following options:

- Set it to 0 before/after the GPU schedule.
- Set it to 0 in a buffer that isn't allocated in shared/local memory.
- If you really need this to happen right where it is, you could try
setting the block coordinate that is equal to 0, 0, 0 in the first block
(or whichever block it is) and wrapping that in a select that has 0 for
true, and reads the block value for false. Unfortunately, this makes the
algorithm no longer independent of the schedule, because the local
coordinates within the block are a function of the block size. This also
*might* still not work because it is a potentially difficult case for the
bounds inference to understand.

**Comment by jaakkotuosa** [comment]
Thanks for the debugging idea. I had not seen this `HL_TARGET=opencl-debug` before, and after changing the lesson to use the env variable I got the debugging info. Indeed, the debug showed that 2949120 bytes (=image width * height * 3) of shared memory is requested (32768 local bytes available).

`select` like this seems to work:
```padded(x, y, c) = select((x == 0) && (y == 0) && (c == 0), 0, padded(x, y, c));```
So no need to refer to actual block coordinates. This way GPU would however be doing `select` for each pixel. To avoid this, maybe the only option is then to do the assignment before/after the GPU schedule, as dsharletg suggested.

Now, update definition `padded(x, y, c) += 1;` seems to work on OpenCL + GPU while `padded(0, y, c) += 1;` or `padded(0, 0, 0) += 1;` does not (both failing with different errors). How would be the Halide-way to handle errors like these? Fail while trying in random fashion as it does now, file a bug on the failures, document the limitiation somewhere?

**Comment by dsharletg** [comment]
Sorry for the delay. I don't have a good answer. I think that Halide's programming model for shared memory on GPUs is pretty tricky to understand completely. I'm not sure if we've documented it anywhere. It won't be an easy thing to explain. The only documentation of this that I'm aware of is here: https://github.com/halide/Halide/blob/master/tutorial/lesson_12_using_the_gpu.cpp#L157 but it is admittedly sparse. 

The challenge with it is that fully documenting this requires documenting Halide's bounds inference process, which, ideally, would not be something that Halide users would need to worry about. It's something we will need to think about long term, but I think this will be a sharp corner for a while.

**Comment by dsharletg** [comment]
This is a difficult thing for us to document. We would ultimately like for
Halide users not to have to be aware of things like bounds inference that
Halide has to do internally, but this particular corner of Halide cannot be
understood without understanding how bounds inference works.

You are right to ask about what to do about this... but I don't think we
are prepared to make a decision on this at the moment. For now, I suggest
handling them ad-hoc (github issue/Halide-dev message), as you did here.
Sorry I don't have a more satisfying answer to this :)

**Comment by gchauras** [comment]
I am writing a lot of GPU schedules and having a good understanding of bounds inference is essential for shared memory and debugging in general. Right now, I look at the codegen and try to backtrack the logic which caused a particular bounds inference.
This is hard as it as, never mind a deliberate effort to obscure bounds inference even further. I would like Halide to be more upfront about how bounds are inferred - an HL_TRACE kind of mode which simply tells me the order in which bounds are inferred. Many a times you have to trick bounds inference into doing what you actually need. I would say power users will eventually need to understand this and even for first timers, it is really frustrating to see failed kernel launches when you don't understand why.

**Comment by dsharletg** [comment]
I agree that understanding bounds inference is necessary to work with Halide, especially when working with GPU schedules using shared memory. There is no intentional obfuscation of bounds inference information, it just generates a *ton* of information, most of which gets simplified away (for optimization purposes, not obfuscation). You can look at what bounds inference is doing by using HL_DEBUG_CODEGEN=2 and looking at the program before most of the bounds inference information is optimized away. The problem is that this is often an overwhelming amount of information. 

So, to clarify my position on this: I'd like to find a more direct way to communicate with the user about  issues like this one (a global index being included in shared tile memory). Bounds inference tells you what the problem is, but it doesn't easily tell you where/why that problem is occurring, without knowing exactly what you are looking for.

**Comment by zvookin** [comment]
We need to do a pass through the debug statements and clean them up. It may
also be better to move to a design where we have specific flags instead of
inclusive numerical levels. That way, one could turn on bounds inference
debugging without turning on a lot of other messages.

It is also important to capture the cases where there is difficulty. Bounds
inference in general is difficult. I've spent plenty of time doing it by
hand and it is tedious and error prone. Generally I find Halide's approach
way easier, but there are tricky cases. The more of these we have captured
in bugs the better chance we have of improving both the inference algorithm
and the user feedback it provides.

I'm guessing one of the issues on the GPU is that we don't have a strong
notion of the limit on certain types on memory. I don't think we have
anything that preflights a kernel's bounds needs against what the GPU it is
running on can provide. That might be a simple thing to add...

-Z-

**Comment by gchauras** [comment]
Big + to the above comment. HL_DEBUG_CODEGEN gets so verbose at 2 that it is practically impossible to find any info. Having descriptive flags will be great.

**Comment by zvookin** [comment]
Oh and some way to do folding/filtering on the debug output ala HTML stmt
files. I'm kind of amazed that there isn't a standard, widely used, library
for higher usability debug logging.

**Comment by dsharletg** [comment]
I also agree, but how? There's no way to know whether a piece of IR is bounds inference related or not. 

If you are simply referring to only dumping out the IR before simplification, and not all the other stuff (e.g. final LLVM IR), that wouldn't be hard to do, but I'm also not sure how much difference that makes. To me, the main verbosity problem is that the lowered IR contains 99% trivial bounds inference lets, and 1% interesting ones. I don't see how to easily filter that down.

> I'm guessing one of the issues on the GPU is that we don't have a strong
notion of the limit on certain types on memory. I don't think we have
anything that preflights a kernel's bounds needs against what the GPU it is
running on can provide. That might be a simple thing to add...

I was actually thinking that we should consider simply not allowing dynamically sized shared/local memory allocations. We already require the tile size to be constant, I'm having a hard time thinking of a legitimate use case of dynamically allocated shared memory. The shared memory allocation size being dynamic seems like a strong predictor of a program accidentally including a global memory index in a tile buffer.

We do have a test that checks that dynamic shared/local memory allocations work, but I don't see how that test represents a real world use case. That particular test would break as soon as the image became larger than a size that would be trivially small for a GPU. If the dynamic allocation were a function of a local/tile parameter to avoid that, it wouldn't be dynamic anymore. @abadams, any thoughts on this?

**Comment by zvookin** [comment]
I'm not sure there is much difference between dynamic sizing on
shared/local memory and different pieces of hardware having different
limits on size. The use case for dynamic would be a LUT that has different
levels of quantization. You can likely make this constant size via e.g.
specialize, but unless removing the flexibility wins, I'd like to keep it.

I'm imagining a model where any memories in the hardware that have a
runtime known fixed size are treated as invisible inputs and we add bounds
checks against them. The goal is to get an error message saying which Func
won't fit and which type of memory it doesn't fit into. Preferably with the
actual bounds of each. There may be a max bound constraint and there are
often max number of dimensions constraints.

-Z-


On Jan 12, 2015, at 10:16 AM, Dillon Sharlet <notifications@github.com>
wrote:

I also agree, but how? There's no way to know whether a piece of IR is
bounds inference related or not.

If you are simply referring to only dumping out the IR before
simplification, and not all the other stuff (e.g. final LLVM IR), that
wouldn't be hard to do, but I'm also not sure how much difference that
makes. To me, the main verbosity problem is that the lowered IR contains
99% trivial bounds inference lets, and 1% interesting ones. I don't see how
to easily filter that down.

I'm guessing one of the issues on the GPU is that we don't have a strong
notion of the limit on certain types on memory. I don't think we have
anything that preflights a kernel's bounds needs against what the GPU it is
running on can provide. That might be a simple thing to add...

I was actually thinking that we should consider simply not allowing
dynamically sized shared/local memory allocations. We already require the
tile size to be constant, I'm having a hard time thinking of a legitimate
use case of dynamically allocated shared memory. The shared memory
allocation size being dynamic seems like a strong predictor of a program
accidentally including a global memory index in a tile buffer.

We do have a test that checks that dynamic shared/local memory allocations
work, but I don't see how that test represents a real world use case. That
particular test would break as soon as the image became larger than a size
that would be trivially small for a GPU. If the dynamic allocation were a
function of a local/tile parameter to avoid that, it wouldn't be dynamic
anymore. @abadams <https://github.com/abadams>, any thoughts on this?

—
Reply to this email directly or view it on GitHub
<https://github.com/halide/Halide/issues/608#issuecomment-69617029>.

**Comment by dsharletg** [comment]
With OpenCL, it will be very difficult to have a known fixed size of shared/local memory. We **could** do this for CUDA by using the compute capability target option to determine the minimum available shared memory.

**Comment by jaakkotuosa** [comment]
Thanks for the comments. It might indeed be the case that there is no simple solution for this particular issue, other than having it written here. I could close this issue, unless someone wants to continue these discussions here?

**Comment by dsharletg** [comment]
I suggest we leave this open. I can't say I expect much change on this in
the very near future, but it's definitely an open issue we should address
at some point.
